Add CMU MoCap Dataset

dnaveenr opened the pull request:
Resolves #3457 

Dataset Request : Add CMU Graphics Lab Motion Capture dataset [#3457](https://github.com/huggingface/datasets/issues/3457)
This PR adds the CMU MoCap Dataset.

The authors didn't respond even after multiple follow ups, so I ended up crawling the website to get categories, subcategories and description information. Some of the subjects do not have category/subcategory/description as well. I am using a subject to categories, subcategories and description map (metadata file).

Currently the loading of the dataset works for "asf/amc" and "avi" formats since they have a single download link. But "c3d" and "mpg" have multiple download links (part archives) and dl_manager.download_and_extract() extracts the files to multiple paths, is there a way to extract these multiple archives into one folder ? Any other way to go about this ?
Any suggestions/inputs on this would be helpful. Thank you.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_4208). All of your documentation changes will be reflected on that endpoint.

dnaveenr commented:
- Updated the readme.
- Added dummy_data.zip and ran the all the tests.
-  The dataset works for "asf/amc" and "avi" formats which have a single download link for the complete dataset. But "c3d" and "mpg" have multiple download links, can we combine and host these links on the Hub since the dataset is free to use ?

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,169 @@
+
+### Dataset Summary
+
+The CMU Graphics Lab Motion Capture Database is a dataset containing motion capture recordings of people performing different actions such as walking, jumping, dancing etc. The actions have been performed by over 140 subjects. There are a total of 2605 trials in 6 categories and 23 subcategories.
```
According to the generated infos file, the number of examples is 112, so we should update the numbers in this section accordingly if that's correct. 

I would expect 144 [subjects](http://mocap.cs.cmu.edu/subjects.php)/rows in the dataset. Is the metadata file missing some subjects? If yes, is there something we can do to get them?

dnaveenr replied:
Yes. The no.of examples are 112, I'll update this section.
The subject numbers go upto 144 (http://mocap.cs.cmu.edu/subjects.php) but some of the subject numbers are missing, so total count of examples come up to 112 itself.

mariosasko replied:
Maybe we can add a note to let the users know that some examples are not present on the website.

mariosasko commented on datasets/cmu_mocap/README.md:
````diff
@@ -0,0 +1,169 @@
+
+## Dataset Structure
+
+### Data Instances
+
+A sample from the CMU MoCap dataset is provided below:
+```
+{
+    'subject_id': 12, 
+    'categories': ['Physical Activities & Sports', 'Locomotion', 'Locomotion', 'Locomotion'], 
+    'subcategories': ['martial arts', 'walking', 'walking', 'walking'], 
+    'descriptions': ['tai chi', 'walk', 'walk', 'walk'], 
+    'amc_files': ['~/all_asfamc/subjects/12/12_04.amc', '~/all_asfamc/subjects/12/12_01.amc', '~/all_asfamc/subjects/12/12_03.amc', '~/all_asfamc/subjects/12/12_02.amc'], 
+    'asf_file': '~/all_asfamc/subjects/12/12.asf',
+}
+```
+Depending on the file format, 'files' field would change. Check the Data Fields section for more information.
+
+### Data Fields
+
+The following fields are common for all file formats of the dataset.
+
+- `subject_id`: Unique ID for each subject in the MoCap dataset.
+- `categories`:  A list containing the category names of all trials.
+- `subcategories`: A list containing the subcategory names of all trials.
+- `descriptions`: A list containing the descriptions of all trials.
+
+According to the file format, we have the following fields:
+- `amc_files`, `c3d_files`, `mpg_files` , `avi_files` : A list containing the path to all files.
+
+The "asf/amc" config has an additional field :
+
+- `asf_file`: ASF file is the skeleton file, in the ASF file a base pose is defined for the skeleton that is the starting point for the motion data. More info can be found [here](https://research.cs.wisc.edu/graphics/Courses/cs-838-1999/Jeff/ASF-AMC.html).
````
Let's expose the other formats via configs rather than via a parameter. Also, let's rename `{format}_files` to `motions` (for the `asf/amc` config this field can be a dictionary).

dnaveenr replied:
Yes, I'll change {format}_files to motions.
Also, may I know why using configs instead of parameter would be a better option ? I'll work on this part.

mariosasko replied:
Parameters are more hidden, and we need an extra section to document them. And configs are shown in the viewer and an error message when the user specifies an unsupported config.

dnaveenr replied:
Added the formats as configs and renamed {format}_files to motions. Updated the readme accordinly.

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,169 @@
+
+### Data Splits
+
+All the data is contained in training set.
```
Let's mention the exact number of examples here.

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,169 @@
+
+### Source Data
+
+#### Initial Data Collection and Normalization
+
+From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
+> The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are mo ...
+
+#### Who are the source language producers?
+
+[More Information Needed]
+
+### Annotations
+
+#### Annotation process
+
+[More Information Needed]
+
+#### Who are the annotators?
+
+[More Information Needed]
+
```
Feel free to include more from this page: http://mocap.cs.cmu.edu/info.php (and even add images)

dnaveenr replied:
Sure. Will do that.

mariosasko commented on datasets/cmu_mocap/cmu_mocap.py:
```diff
@@ -0,0 +1,288 @@
+import datasets
+
+
+_CITATION = """\
+
+"""
```
```suggestion
_CITATION = "N/A"
```

mariosasko commented on datasets/cmu_mocap/cmu_mocap.py:
```diff
@@ -0,0 +1,288 @@
+_HOMEPAGE = "http://mocap.cs.cmu.edu/"
+
+# TODO: Add the licence for the dataset here if you can find it
+_LICENSE = ""
```
```suggestion
_LICENSE = "This data is free for use in research projects. You may include this data in commercially-sold products, but you may not resell this data directly, even in converted form."
```

mariosasko reviewed (commented):
Thanks for working on this! I've left some comments.

It's also important to add a `Dataset Preprocessing` section to show how to process data and to mention specifics of each format. We can use [this](https://github.com/CalciferZh/AMCParser) project to parse asf/amc, [this](https://c3d.readthedocs.io/en/stable/) one for cv3d and [this one](https://github.com/dmlc/decord) for avi/mpg.

dnaveenr commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,169 @@
+- monolingual
+pretty_name: CMU MoCap
+size_categories:
+- 1K<n<10K
```
Yes. The dataset has around 144 subjects, but I looked at the total number of trials present which came upto 2514.

mariosasko replied:
This number should match the number of examples in the dataset, and an example in this dataset corresponds to a subject and its sequence of trials.

dnaveenr replied:
Got it. This makes sense.

dnaveenr commented:
c3d" and "mpg" have multiple download links (part archives) and dl_manager.download_and_extract() extracts the files to multiple paths, is there a way to extract these multiple archives into one folder ? Any other way to go about this ?
Can we combine and host these links on the Hub since the dataset is free to use ?

mariosasko commented:
> "c3d" and "mpg" have multiple download links (part archives) and dl_manager.download_and_extract() extracts the files to multiple paths, is there a way to extract these multiple archives into one folder ? Any other way to go about this ?

We store downloaded data under `~/.cache/huggingface/datasets/downloads`, so these downloads are "hidden" and won't clutter one's filesystem in an "obvious way".

dnaveenr commented:
> We store downloaded data under ~/.cache/huggingface/datasets/downloads (by default), so these downloads are "hidden" and won't clutter one's filesystem in an "obvious way".

Yes, the filesystem won't be clustered, but the problem is processing the dataset becomes cumbersome. For eg, for the c3d format has 5 part-downloads, so the folders will be as follows : 
```
['~/.cache/huggingface/datasets/downloads/extracted/0e6bf028f490bf18c23ce572d1437c4ef32a74f630e33c26a806250d35cfcdd1', '~/.cache/huggingface/datasets/downloads/extracted/1b44fc5c7a6e031c904545422d449fd964f8ee795b9d1dcb0b6a76d03b50ebe6', '~/.cache/huggingface/datasets/downloads/extracted/137595188e96187c24ce1aa5c78200c7f78816fbd9d6c62354c01b3e6ec550c7', '~/.cache/huggingface/datasets/downloads/extracted/6c0c893e435f36fd79aa0f199f58fe16f01985f039644a7cb094a8c43a15ffd4', '~/.cache/huggingface/datasets/downloads/extracted/45e4703354cbc975e6add66f1b17b716c882b56f44575b033c5926aa5fcfb17f']
```
Each of these folders have a given set of subjects, so we'll be need to write extra code to fetch data from each of these folders, and the mpg format has 12 part-downloads which will lead to 12 folders having certain set of subjects, so it is cumbersome to process them.

dnaveenr commented:
I have added all the changes that were suggested. We just need to handle the multi-part download for c3d and mpg formats. Easiest way would be to have just one zip for these formats.

mariosasko commented:
But we can handle this with a simple mapping that stores the ranges of ids (for each config), no? And an actual file path is not important during processing.

dnaveenr commented:
I have added code to handle c3d, mpg formats as well.  The data for the mpg format seems incomplete as it contains only 53 rows. I have added a note regarding this in the Data Splits section.

dnaveenr commented:
The real data test works fine. But I get the following error for dummy data test : 
```
            try:
                # Prepare split will record examples associated to the split
                self._prepare_split(split_generator, **prepare_split_kwargs)
            except OSError as e:
                raise OSError(
                    "Cannot find data file. "
                    + (self.manual_download_instructions or "")
                    + "\nOriginal error:\n"
                    + str(e)
>               ) from None
E               OSError: Cannot find data file. 
E               Original error:
E               [Errno 21] Is a directory: '/tmp/tmpbmcvs0cl/extracted/0699e42b1104339701944206ae06708ac8296d937907280da3e4a927718f7df6/dummy_data'

src/datasets/builder.py:709: OSError
```
The dummy data looks fine to me, not sure what the issue is.

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+### Data Fields
+
+The following fields are common for all file formats of the dataset.
+
+- `subject_id`: Unique ID for each subject in the MoCap dataset.
+- `categories`:  A list containing the category names of all trials.
+- `subcategories`: A list containing the subcategory names of all trials.
+- `descriptions`: A list containing the descriptions of all trials.
+
+The c3d, mpg and avi formats have: 
+- `motions` : A list containing the path to all motion files.
+
+The "asf/amc" config has an additional field :
+- `motions`: {
+        `amc_files`: A list containing the path to all motion files, 
+        `skeleton_file`: path to the ASF file,
+    }
+
+    It is a dictionary containing the motion files and skeleton file.
```
```suggestion
The following fields are common for all file formats of the dataset.

- `subject_id`: Unique ID for each subject in the MoCap dataset.
- `categories`:  A list containing the category names of all trials.
- `subcategories`: A list containing the subcategory names of all trials.
- `descriptions`: A list containing the descriptions of all trials.

The subject's motion trials in the `asf-amc` format are represented as follows:
- `motions`:
  - `asf`: A list containing the paths to all ASF motion files
  - `amc`:  A list containing the paths to all AMC motion files

In the `c3d`, `mpg` and `avi` formats, this field has the following structure:-
- `motions`: A list containing the path to all motion files.
```

PS: Please also update the features accordingly

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+    It is a dictionary containing the motion files and skeleton file.
+
+ASF file is the skeleton file, in the ASF file a base pose is defined for the skeleton that is the starting point for the motion data. More info can be found [here](https://research.cs.wisc.edu/graphics/Courses/cs-838-1999/Jeff/ASF-AMC.html).
```
It's better to have this info in the Dataset Preprocessing section

dnaveenr replied:
Moved to Dataset Preprocessing section.

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+languages:
+- en
+licenses:
+- unknown
```
```suggestion
- other
```

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+### Data Splits
+
+All the data is contained in training set. 
+If you consider the "asf/amc" format of the dataset, there are a total of :
+- 112 subjects/examples
+- Total number of trials(.amc files) in the dataset are 2514.
+- There are 112 asf files for each subject.
+
+Note :- 
+Each format contains the following rows :
+- asf-amc : num_rows - 112,
+- c3d :  num_rows - 109,
+- mpg : num_rows - 53,
+- avi : num_rows - 106.
```
```suggestion
All the data is contained in training set. 

Not all subjects were converted from the initial `asf-amc` format to the destination formats, so depending on the format, the number of examples varies:
<MARKDOWN TABLE>
```

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+#### Initial Data Collection and Normalization
+
+From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
+> The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are mo ...
+
+This 3D data can be used in two ways by you:
+
+- Marker positions You can be handed a file of 3D marker positions, a .c3d . This file is relatively clean - i.e., Marker "A" should be labeled Marker "A" throughout the motion. But it is your responsibility to figure out what "A" means and how it relates to the other markers.
+- Skeleton movement Data will be handed to you as either a .vsk/.v pair or .asf/.amc pair (more on that later). The former element of the pair describes the skeleton and its joints: their connections, lengths, degrees of freedom (free, ball and socket, 2 hinges, hinge, rigid), and mathematical transformations. The latter element of the pair contains the movement data. Notes: If a subject/object w ...
```
```suggestion
From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
> The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are moved closer to capture a smaller space with higher resolution. To capture something, small grey markers are placed on it. Humans wear a black jumpsuit and have 41 markers taped on. The Vicon cameras see the markers in infra-red. The images that the various cameras pick up are triangulated to get 3D data.
>
This 3D data can be used in two ways by you:
>
- Marker positions You can be handed a file of 3D marker positions, a .c3d . This file is relatively clean - i.e., Marker "A" should be labeled Marker "A" throughout the motion. But it is your responsibility to figure out what "A" means and how it relates to the other markers.
- Skeleton movement Data will be handed to you as either a .vsk/.v pair or .asf/.amc pair (more on that later). The former element of the pair describes the skeleton and its joints: their connections, lengths, degrees of freedom (free, ball and socket, 2 hinges, hinge, rigid), and mathematical transformations. The latter element of the pair contains the movement data. Notes: If a subject/object was captured in multiple clips, you will be handed several .v's or .amc's. Also, something like a hamburger turner, if that's what you're capturing, can have a "skeleton" - even if it's one bone long.
```

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+### Dataset Curators
+
+[Needs More Information]
```
```suggestion
Jessica K. Hodgins
```

mariosasko commented on datasets/cmu_mocap/cmu_mocap.py:
```diff
@@ -0,0 +1,300 @@
+        ),
+    ]
+
+    BUILDER_CONFIG_CLASS = CmuMocapConfig
+
```
`asf-amc` was the initial format and has the most data, so let's make it the default config:
```suggestion
    BUILDER_CONFIG_CLASS = CmuMocapConfig
    DEFAULT_CONFIG_NAME = "asf-amc"
```

mariosasko reviewed (commented):
A few more changes and we can merge.

PS: We try to occupy less than 500 KB for dummy data per dataset. Is there something you can to reduce the size?

dnaveenr commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,252 @@
+
+Note : The dataset has subjects numbered upto 144 [here](http://mocap.cs.cmu.edu/subjects.php) but some of the subject numbers are missing, so the total examples are around 112.
+
+### Dataset Preprocessing
+
+The dataset contains motions in the following formats :
+
+- asf-amc : The ASF file (Acclaim Skeleton File) is a skeleton file . AMC files (Acclaim Motion Capture data) are the motion files. Most of the time a single skeleton works for many different motions and rather than storing the same skeleton in each of the motion files, thus for each subject there is one ASF file along with AMC files for each motion.
+
+The AMC/ASF files can be parsed using [AMCParser library](https://github.com/CalciferZh/AMCParser). You can check the Usage section in this libary for more information.
+
+- c3d : The C3D format stores 3D coordinate information, analog data and associated information used in 3D motion data capture and subsequent analysis operations. 
+
+You can use [py-c3d library](https://c3d.readthedocs.io/en/stable/) for reading and writing C3D binary files.
+
+- avi and mpg : AVI (Audio Video Interleave) has been the long standing format to save and deliver movies and other video files. An MPG file is a common video file that uses a digital video format standardized by the Moving Picture Experts Group (MPEG)
+
+You can use a video loader library such as [decord](https://github.com/dmlc/decord) to load and process these files.
```
Okay. Will try adding the preprocessing code for each format.

dnaveenr commented:
- Reduced the dummy_data size.
- Added sample dataset preprocessing code, it is not complete though.
- Added all changes suggested.

Let me know if anything else is required. Thank you. :)

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,307 @@
+All the data is contained in training set. 
+
+Not all subjects were converted from the initial `asf-amc` format to the destination formats, so depending on the format, the number of examples varies:
+<MARKDOWN TABLE>
```
What I meant by `<MARKDOWN TABLE>` is a markdown table with the size of each dataset (a table similar to [this one](https://huggingface.co/datasets/imdb#data-splits)) should be placed here.

mariosasko commented on datasets/cmu_mocap/README.md:
```diff
@@ -0,0 +1,307 @@
+The following fields are common for all file formats of the dataset.
+
+- `subject_id`: Unique ID for each subject in the MoCap dataset.
+- `categories`:  A list containing the category names of all trials.
+- `subcategories`: A list containing the subcategory names of all trials.
```
Feel free to add the label mapping here.

mariosasko commented on datasets/cmu_mocap/cmu_mocap.py:
```diff
@@ -0,0 +1,301 @@
+
+        return subject_paths
+
+    def _generate_examples(self, data_paths, metadata_path, split):
```
The URLs support streaming from what I see, so it would be cool to make the dataset streamable. Let me know if you need help with that.

mariosasko reviewed (commented):
Good job with the dummy data!

albertvillanova commented:
Thanks for your contribution, @dnaveenr.

We are removing the dataset scripts from this GitHub repo and moving them to the Hugging Face Hub: https://huggingface.co/datasets

We would suggest you create this dataset there. Please, feel free to tell us if you need some help.
